### ISSUE TITLE
Measure user usage of the Cribbage Trainer automatically deployed to GitHub Pages

### ISSUE THREAD

**Issue by markafitzgerald1** [markafitzgerald1/cribbage-trainer#21]
At present there is no available measure of how much traffic the automatically deployed to [GitHub Pages version of Cribbage Trainer](https://markafitzgerald1.github.io/cribbage-trainer/) is getting, if any.

Add some sort of tracker so that once the application is getting traffic, it's application development and roadmap can be adjusted and fine-tuned based on its global usage information and trends.

**Comment by markafitzgerald1** [comment]
I'm currently investigating whether [Google Analytics](https://analytics.google.com/analytics/web/provision/?authuser=0#/provision) will suffice as a starting point to provide basic user usage info.

**Comment by markafitzgerald1** [comment]
Consider adding per user action tracking as well for:
- [ ] discard selection, and
- [ ] discard deselection

in order to measure how much users use the app per page load. (#6 will affect these analytics, allowing users to stay on the same page without reloading for multiple hands.)

**Comment by markafitzgerald1** [comment]
The [Google Analytics Terms and Conditions](https://marketingplatform.google.com/about/analytics/terms/us/) requires a Privacy Policy and user consent to Google Analytics cookies and/or local storage in section 7. Privacy:

> You will have and abide by an appropriate Privacy Policy and will comply with all applicable laws, policies, and regulations relating to the collection of information from Users. You must post a Privacy Policy and that Privacy Policy must provide notice of Your use of cookies, identifiers for mobile devices (e.g., Android Advertising Identifier or Advertising Identifier for iOS) or similar technology used to collect data. You must disclose the use of Google Analytics, and how it collects and processes data. This can be done by displaying a prominent link to the site "How Google uses information from sites or apps that use our services", (located at [www.google.com/policies/privacy/partners/](https://policies.google.com/technologies/partner-sites), or any other URL that Google may provide from time to time). You will use commercially reasonable efforts to ensure that a User is provided with clear and comprehensive information about, and consents to, the storing and accessing of cookies or other information on the User’s device where such activity occurs in connection with the Service and where providing such information and obtaining such consent is required by law.

Working on that now.

- [ ] Create and add a Privacy Policy
- [ ] Create and add a user consent to the use of Google Analytics user interface

**Comment by markafitzgerald1** [comment]
Adding [Storybook](https://storybook.js.org/) to more effectively iterate on the Google Analytics consent widget before incorporating it into the main user interface.

- [ ] Add Storybook
- [ ] Create Analytics consent widget

**Comment by markafitzgerald1** [comment]
[Branch](https://github.com/markafitzgerald1/cribbage-trainer/tree/21-measure-cribbage-trainer-github-pages-usage) and [draft PR](https://github.com/markafitzgerald1/cribbage-trainer/pull/27) created with dependencies updated and deploy to production disabled for non-`main` branches.

GitHub repository settings updated to require PR passing build and test check before merge to `main` to avoid any future production outages due to GitHub Actions detectable issues.

**Comment by markafitzgerald1** [comment]
Storybook [does not appear to support Parcel](https://github.com/storybookjs/storybook/tree/next/code/frameworks). Default Storybook instructions only show how to use it with Webpack and Vite. Vite seems to be more future-proof and its use of ECMAScript modules very likely works with the evergreen browsers this project is targeting. Perhaps consider a move to [Vite](https://vitejs.dev/) via their [migration guide from Parcel](https://gock.net/blog/2023/convert-react-svelte-project-from-parcel-to-vite/)) as a supported path to using Storybook? There's no obvious sign of Parcel support coming to Storybook and a move to Vite would buy some local development [performance gains](https://storybook.js.org/blog/storybook-for-vite/). Also, Storybook based UI component development [looks rather awesome](https://storybook.js.org/blog/storybook-for-vite/)!

- [ ] Prototype migration from Parcel to Vite to enable use of Storybook (to enable ease of development of this new UI component in Cribbage Trainer).

**Comment by markafitzgerald1** [comment]
TIL: The GitHub Pages Action has a feature in development to:
> [deploy a pull request as a GitHub Pages preview site](https://github.com/marketplace/actions/deploy-github-pages-site#inputs-)

The feature is still in non-public alpha, but let's monitor to enable the possibility of deploys to a preview site from PRs.

**Comment by markafitzgerald1** [comment]
Install and use of Vite followed by disuse and uninstall of Parcel may be easier and another than using a third party migration guide.

**Comment by markafitzgerald1** [comment]
Replaced Parcel with Vite in 633c7d6bd66cf2c6898183d46d2e39082ad202b2.

**Comment by markafitzgerald1** [comment]
The `Package` production dependency looks like an accidental error. Please remove in this PR if it is the next one to merge into `main`.

**Comment by markafitzgerald1** [comment]
Storybook with Card stories added in 4281a13c95087af80a10579c59fde3c097b4cbec.

**Comment by markafitzgerald1** [comment]
`Package` dependency removed in 8aaa89a6d56748e9473c263a326f17bfa1b374e7.

**Comment by markafitzgerald1** [comment]
Storybook Test Runner incorporated in local and GitHub Actions checks in 81f6df8.

**Comment by markafitzgerald1** [comment]
PR branch getting [too big](https://github.com/markafitzgerald1/cribbage-trainer/pull/27/files): 9,654 lines added and 3,421 lines removed. I think I'll start merging the supporting changes for the resolution of this issue into `main` one at a time to reduce PR size and branch duration (currently 25 days old) in order to keep non-merged work in progress size more manageable and understandable.

**Comment by markafitzgerald1** [comment]
Automated local serving of Storybook for Storybook tests in the [21-measure-cribbage-trainer-github-pages-usage](https://github.com/markafitzgerald1/cribbage-trainer/tree/21-measure-cribbage-trainer-github-pages-usage) branch to automate one sometimes needed local development separate step.

**Comment by markafitzgerald1** [comment]
Adding Storybook well will likely require modifying how the existing components work as stories are added.

TODO:
- [ ] Change `Card` to only fire events on `change` - not actually modify app state.
- [ ] Add stories for:
  - [ ] `Hand`,
  - [ ] `SortOrderInput`,
  - [ ] `SortableHand`,
  - [ ] `PossibleHandCard`,
  - [ ] `PossibleHand`,
  - [ ] `Calculation`,
  - [ ] `Calculations`, and
  - [ ] `Trainer`

**Comment by markafitzgerald1** [comment]
822cfae pushed to [this branch](https://github.com/markafitzgerald1/cribbage-trainer/tree/21-measure-cribbage-trainer-github-pages-usage) and [draft PR](https://github.com/markafitzgerald1/cribbage-trainer/pull/27) to simplify the Card component to fire events instead of directly modifying state via parent component overall hand state hook calls.

**Comment by markafitzgerald1** [comment]
`CardLabel` stories added in 127963b8a34fc7f9f2b33ca17c209f879a3f18c9.

**Comment by markafitzgerald1** [comment]
Probably `PossibleHand` is next to get Storybook stories, or perhaps `Hand`.

**Comment by markafitzgerald1** [comment]
PR has gotten too big. Will merge it into `main`, then continue on this issue using trunk-based development to reduce cognitive load (only one branch) and keep the tip of development production tested.

Issue resolution strategy remains the same: supporting technical infrastructure before feature + issue remains open until both technical support and the feature are complete with minimal technical debt, where all technical debt is captured in Issues.

**Comment by markafitzgerald1** [comment]
TODO:
- [ ] merge and close PR

**Comment by markafitzgerald1** [comment]
First three commits of the PR (#27) now exist in `main`:
- move to Node.js 20,
- restrict of deploy to occur from `main` only, and
- move from Parcel to Vite (to enable add of Storybook).

Next: add Storybook in `main`.

**Comment by markafitzgerald1** [comment]
Next three commits of the PR (#27) now exist in `main`:
* add Storybook and Card stories,
* ensure test screenshots up to date on Linux, and
* remove unused `Package` dependency.

Next: use the Storybook [Test Runner](https://storybook.js.org/docs/react/writing-tests/test-runner).

**Comment by markafitzgerald1** [comment]
TODO:
- [ ] Publish built storybook to GitHub Pages alongside SPA if readily possible.

**Comment by markafitzgerald1** [comment]
[🐞 fix(Card): Minimize per components props](https://github.com/markafitzgerald1/cribbage-trainer/pull/27/commits/822cfaee8c29f1f3349fc94ff8f76594e412d9fa) is now in `main`, but it contains the regression where clicking between the card border and the card text does not result in card select/deselect happening. I believe it's fixed in the last commit - will try merging that commit in now.

**Comment by markafitzgerald1** [comment]
Pausing work on this until some known bugs get fixed: #33, #31, #12.

**Comment by markafitzgerald1** [comment]
#31 closed.
#12 deferred until after #26, which will happen naturally a little after this Issue (#21).
#33 in progress and will get closed during next PR or two of this Issue (#21).

Resuming work on this Issue (#21), which will naturally result in #33 closing out before #21 anyways.

**Comment by markafitzgerald1** [comment]
`PossibleHand` feels like the next candidate component for which Storybook stories should be added.

**Comment by markafitzgerald1** [comment]
`PossibleHand` stories added in #44.

**Comment by markafitzgerald1** [comment]
`ScoredPossibleKeepDiscard` stories added in #48.

**Comment by markafitzgerald1** [comment]
`Hand` stories added in https://github.com/markafitzgerald1/cribbage-trainer/pull/58.

**Comment by markafitzgerald1** [comment]
`SortOrderInput` stories added in #59.

**Comment by markafitzgerald1** [comment]
TODO inherited from #70:
- [ ] Add a Storybook interaction test to prevent this from re-occurring once `SortableHand` Stortbook stories exist (via #21 work).

**Comment by markafitzgerald1** [comment]
`SortableHand` stories added in #78.

**Comment by markafitzgerald1** [comment]
`ScoredPossibleKeepDiscards` stories added in https://github.com/markafitzgerald1/cribbage-trainer/pull/79.

**Comment by markafitzgerald1** [comment]
`Trainer` story added in #80.

**Comment by markafitzgerald1** [comment]
Storybook static site now automatically published to https://markafitzgerald1.github.io/cribbage-trainer/storybook/ with #81 and #82.

**Comment by markafitzgerald1** [comment]
Minor glitch with Docs view of Trainer Story acceptable for now - can fix it in https://github.com/markafitzgerald1/cribbage-trainer/issues/83 later.

**Comment by markafitzgerald1** [comment]
Next up: review https://storybook.js.org/docs/writing-tests/interaction-testing before adding the regression test that was preventing sort from working in #70 before starting to create the analytics consent widget.

**Comment by markafitzgerald1** [comment]
Locally wrote a Storybook Interaction Test of hand sort in Trainer. It warns in the browser console, but passes `npx test-storybook`.

A Storybook test may be required here as the development version of the component is used in Storybook, and that version of the component does not fail to work during test execution.

TODO:
- [ ] Write Playwright test of changing sort order to see if that would catch #70 if it happened again.

**Comment by markafitzgerald1** [comment]
Confirmed that `act()` in code + Playwright test using production preview server catches the use of `act()` due to UI not changing or looking as expected after click on different sort order radio button.

**Comment by markafitzgerald1** [comment]
Bottom line: both the Storybook Interaction test and the Playwright test can catch the issue when run against production app servers.

**Comment by markafitzgerald1** [comment]
Regression test for #70 added in https://github.com/markafitzgerald1/cribbage-trainer/pull/86.

Next up: analytics consent UI create!

**Comment by markafitzgerald1** [comment]
Privacy Policy component created and added to Storybook but not added to main user interface in https://github.com/markafitzgerald1/cribbage-trainer/pull/87.

**Comment by markafitzgerald1** [comment]
The privacy policy was created with [Google Bard](https://bard.google.com/) using the parts of [Google Analytics Terms of Service](https://marketingplatform.google.com/about/analytics/terms/us/) Section 7, Privacy that require a Privacy Policy from using applications as input.

**Comment by markafitzgerald1** [comment]
One missing TODO item once the Privacy Policy and a working user consent are on the GitHub Pages deployed site:
- [ ] Use Google Analytics in the GitHub Pages deployed version of Cribbage Trainer when consented to by the user.

**Comment by markafitzgerald1** [comment]
There are plenty of user consent to the user of analytics examples on the Internet that can be used to form one for this app. See https://www.google.com/search?q=analytics+user+consent+dialog+examples for some of them.

**Comment by markafitzgerald1** [comment]
Analytics Consent dialog with Accept and Decline buttons progressing nicely in local development. Close to ready to push as a Storybook only component before integration work begins.

**Comment by markafitzgerald1** [comment]
Pushed Analytics Consent dialog in 68d8854eb06b12cd9257a725057622feb8553e4e.

**Comment by markafitzgerald1** [comment]
Analytics Consent component now in `main`. Working on add of that and Privacy Policy to main UI now.

**Comment by markafitzgerald1** [comment]
[Mixpanel](https://mixpanel.com/pricing/) and [Clicky](https://clicky.com/) also offer free analytics which would provide page load counts of Cribbage Trainer. Perhaps try those out as well?

TODO:

- [ ] Try out Mixpanel.
- [ ] Try out Clicky.

**Comment by markafitzgerald1** [comment]
All dependencies except `eslint` updated to latest. ESLint 9 can be successfully installed if `typescript-eslint` is upgraded to version `8.0.0-alpha-10` or greater now that https://github.com/typescript-eslint/typescript-eslint/issues/8211 is Closed, but other dependencies like `eslint-plugin-jest` do not seem to be happy to work with `@typescript-eslint/eslint-plugin@^8`.

I'll defer on upgrading to ESLint 9 for now, waiting until at least a non-alpha, `@latest` release of `typescript-eslint` exists which works with ESLint 9. By that time, hopefully all transitive dependencies also work with ESLint 9.

**Comment by markafitzgerald1** [comment]
I had a quick look over Mixpanel and Clicky as alternatives to using Google Analytics, and concluded that because Google Analytics is suitable for both now (small-scale, growing app) and the future (during the growth phase), I'll stick with Google Analytics as the initial analytics provider for Cribbage Trainer.

**Comment by markafitzgerald1** [comment]
All dependencies except `eslint` updated to latest. `eslint` upgrade to version 9 no longer blocked on https://github.com/typescript-eslint/typescript-eslint/issues/8211, but is still blocked on https://github.com/jsx-eslint/eslint-plugin-jsx-a11y/issues/978.

**Comment by markafitzgerald1** [comment]
"Add the user consent component to the main application user interface" TODO almost ready for push to branch: functional code is complete, style is complete, unit test coverage is almost complete.

**Comment by markafitzgerald1** [comment]
"Add the user consent component to the main application user interface" TODO added in draft PR https://github.com/markafitzgerald1/cribbage-trainer/pull/151. 

Will still need "Add the Privacy Policy component to the main application user interface" and "Use Google Analytics in the GitHub Pages deployed version of Cribbage Trainer when consented to by the user" TODOs done before that PR can be merged and deployed, as all three parts are required together for the production app to make sense.

**Comment by markafitzgerald1** [comment]
Privacy Policy improved upon in #151. TODO: add Policy to UI from both analytics consent dialog _and_ main app UI.

**Comment by markafitzgerald1** [comment]
TODO: review policy for any missing bits (e.g. meaning of local storage) and possibly simplifications (e.g. data retention period as it is the Google Analytics (GA) default, cookies section when cookies only used for GA, third party data sharing section when not done or currently planned to be used).

**Comment by markafitzgerald1** [comment]
Local storage explanation for non-technical readers added to Privacy Policy in #151.

**Comment by markafitzgerald1** [comment]
Cookies section of Privacy Policy simplified in #151.

**Comment by markafitzgerald1** [comment]
Data Sharing section of Privacy Policy simplified in #151.

**Comment by markafitzgerald1** [comment]
Privacy Policy content improvements complete in #151.

**Comment by markafitzgerald1** [comment]
Privacy policy added and linked in #151 commit [c88a91e](https://github.com/markafitzgerald1/cribbage-trainer/pull/151/commits/c88a91ea2052798699d4fad8c690fd8f6000c3e5).

Google Analytics 4 loaded with consent mode in #151 in [bce7033](https://github.com/markafitzgerald1/cribbage-trainer/pull/151/commits/bce70331af54cf10e3ccb1ef2eeb8bc3a02093a1).

#151 merged into the `main` branch and deployed to production as of 2024-09-22T22:36:04-04.

**Comment by markafitzgerald1** [comment]
Google Analytics says "Your Google tag wasn't detected on your website." after running its test.

I concur: no Google Analytics getting loaded in production. Why not?

**Comment by markafitzgerald1** [comment]
2024-09-22T23:53:43-04: Google Analytics 4 tag now detected on the production Cribbage Trainer site. Fix was to use the `vars` context: https://docs.github.com/en/actions/writing-workflows/choosing-what-your-workflow-does/accessing-contextual-information-about-workflow-runs#vars-context.

**Comment by markafitzgerald1** [comment]
However, my Firefox client is not loading the GA script. Investigation continues...

**Comment by markafitzgerald1** [comment]
uBlock was blocking Google Tag Manager in Firefox. Reconfiguration of Firefox uBlock to allow Google Tag Manager on https://markafitzgerald1.github.io/cribbage-trainer/ resolved that issue.

Google Tag Manager also loading in my Chrome.

No additional traffic sent that I can see on accept of analytics.

No traffic showing from my Firefox or my Chrome in the Google Analytics Realtime view or Home Page (Dashboard).

**Comment by markafitzgerald1** [comment]
Adding the Google Analytics advised script tags:
```
<!-- Google tag (gtag.js) -->
<script async src="https://www.googletagmanager.com/gtag/js?id=G-XXXXXXXXXX"></script>
<script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'G-XXXXXXXXXX');
</script>
```
directly to `<head>` results in an event being sent to Google Analytics.

Now that we have a known working state, can compare against current `main` / production code to figure out why the above works and production does not!

**Comment by markafitzgerald1** [comment]
The issue was that the Google Analytics gtag function used the `arguments` [array-like object](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/arguments) while the production code was using a `...args` [rest parameter](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/rest_parameters).

The fix was to go back to using the `arguments` array-like object in the `gtag` function.

**Comment by markafitzgerald1** [comment]
Issue fixed in https://github.com/markafitzgerald1/cribbage-trainer/pull/161, deployed 20240924T001458-04 and confirmed working T001902-04.

**Comment by markafitzgerald1** [comment]
Event tracking looks correct in Google Analytics. Will leave this issue open at least a day or two to confirm analytics are gathered for users that are not me. :)

**Comment by markafitzgerald1** [comment]
I'm not certain that analytics events on discard selection and deselection is useful at present. Will ponder before explicitly deciding whether to do that in this Issue or not. (Base GA tracking may be enough for current needs and to close this Issue.)

**Comment by markafitzgerald1** [comment]
It may be more effective to not track any custom app events in this issue. Instead, the events to up track can be selected in order to help decide whether and when to build new features, and to measure the impacts of new features.

**Comment by markafitzgerald1** [comment]
Analytics still seems to be working.

Ideas after this issue closes:
- [ ] Get Google to index production app site - once app is more useful.
- [ ] Buy a domain name and point it at the GitHub Pages site - again, once the app is more useful.

**Comment by markafitzgerald1** [comment]
Won't ask Google to index (#22) just yet - app needs to be more of a minimum lovable product (MLP) first. Same for domain name (#163).

**Comment by markafitzgerald1** [comment]
#164 opened to get some sort of measure of usage by Google Analytics blocking users.

**Comment by markafitzgerald1** [comment]
All done!
